### PR TITLE
EZP-24994: Fix "Default Location" for REST API

### PR DESCRIPTION
**Pull request by nicolas-bastien**
https://jira.ez.no/browse/EZP-24994

### CODE REVIEW

**Comment by nicolas-bastien** [comment]
@bdunogier ready for review

File: eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationListProcessor.php [hunk 1]
```diff
@@ -39,4 +39,17 @@ function ($contentId) {
 
         return $outgoingValueHash;
     }
+
+    public function postProcessFieldSettingsHash($outgoingSettingsHash)
+    {
+        $outgoingSettingsHash = parent::postProcessFieldSettingsHash($outgoingSettingsHash);
+
+        if (isset($outgoingSettingsHash['selectionDefaultLocation'])) {
+            $outgoingSettingsHash['selectionDefaultLocationHref'] = $this->mapToContentHref(
```
Comment (bdunogier) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationListProcessor.php]: You can't use `mapToContentHref` here: the id we are mapping is a location id, not a content id. The url needs to be generated using the `ezpublish_rest_loadLocation` route, with a `locationId` parameter.

**Comment by nicolas-bastien** [comment]
@bdunogier done

File: eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationProcessor.php [hunk 2]
```diff
@@ -37,7 +37,7 @@ public function postProcessFieldSettingsHash($outgoingSettingsHash)
         $outgoingSettingsHash = parent::postProcessFieldSettingsHash($outgoingSettingsHash);
 
         if (isset($outgoingSettingsHash['selectionRoot'])) {
-            $outgoingSettingsHash['selectionRootHref'] = $this->mapToContentHref(
+            $outgoingSettingsHash['selectionRootHref'] = $this->mapToLocationHref(
                 $outgoingSettingsHash['selectionRoot']
```
Comment (bdunogier) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationProcessor.php]: We need to check what's stored as the selectionRoot value. If a location path is stored (1/2/42), then we're good. not good, and we need to load the matching location object, and get its actual path (1/2/42).

The easiest is to load a content item that has a relation field with this selectionRoot set, and see what comes out of this implementation. If it's a path like location/42, the we're not good:

`curl http://localhost:8000/api/ezp/v2/content/objects/<id> -H 'Accept: application/vnd.ez.api.Content+xml'`

Comment (dpobel) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationProcessor.php]: selectionRoot only contains a location id, so you are right `selectionRootHref` won't have the expected value

File: eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationProcessor.php [hunk 3]
```diff
@@ -31,4 +31,17 @@ public function postProcessValueHash($outgoingValueHash)
 
         return $outgoingValueHash;
     }
+
+    public function postProcessFieldSettingsHash($outgoingSettingsHash)
+    {
+        $outgoingSettingsHash = parent::postProcessFieldSettingsHash($outgoingSettingsHash);
+
+        if (isset($outgoingSettingsHash['selectionRoot'])) {
```
Comment (dpobel) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/RelationProcessor.php]: `isset` does not seem to be enough, it seems like `selectionRoot` can be set with `''`
I got the following output with your patch https://gist.github.com/dpobel/f13327b30dadaed93c2cff7e449f8934

File: eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php [hunk 4]
```diff
@@ -16,21 +17,55 @@
      */
     private $router;
 
+    /**
+     * @var \eZ\Publish\Core\Repository\LocationService
```
Comment (bdunogier) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php]: TypeHint on `eZ\Publish\API\Repository\LocationService`, not `Core`.

File: eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php [hunk 5]
```diff
@@ -16,21 +17,55 @@
...
         return $this->router->generate('ezpublish_rest_loadContent', ['contentId' => $contentId]);
     }
 
+    /**
+     * @param  int $locationId
+     * @return string
+     */
+    public function mapToLocationHref($locationId)
+    {
+        return $this->router->generate('ezpublish_rest_loadLocation', [
```
Comment (dpobel) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php]: it seems this generates `http://ez.loc/api/ezp/v2/content/locations//1/43/` (notice the double `/`) while this should be `http://ez.loc/api/ezp/v2/content/locations/1/43/` (with the `//` the REST API answers with a 404)

Comment (dpobel) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php]: The ending `/` so actually in the output we should even get `http://ez.loc/api/ezp/v2/content/locations/1/43`

**Comment by nicolas-bastien** [comment]
@dpobel @bdunogier done

File: eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php (hunk 5 above)
Comment (bdunogier) [inline eZ/Publish/Core/REST/Common/FieldTypeProcessor/BaseRelationProcessor.php]: You should check what is done by REST when generating links to locations. Look in `REST\Server\Output\ValueObjectVisitor`.

**Comment by bdunogier** [comment]
We will squash on merge, don't bother rebasing, @nicolas-bastien.

Good job.

**Comment by miguelcleverti** [comment]
PR approved by QA

**Pull request merged**